### ISSUE TITLE
[Feature request] tf.data.Dataset sort and skip buckets

### ISSUE THREAD

**Issue by georgesterpu** [tensorflow/tensorflow#14250]
Hi
It would be useful to sort the variable length inputs by their lengths in order to accelerate the training process. However I cannot find this functionality yet.

In [[1]](https://github.com/tensorflow/tensorflow/issues/7951#issuecomment-308789560), @guillaumekln already suggested something similar through his code snippet, yet the requested feature was batching inputs of similar length together, regardless of the processing order of the batches, and the solution of @mrry in [[2]](https://github.com/tensorflow/tensorflow/issues/7951#issuecomment-326098305) using `group_by_window()` addressed this request just fine. _First question_: Would it be possible to make the iterator return the batches in the ascending order of their ids (given by `key_func`), while maintaining the shuffling operation applied before batching?

Additionally, I would like to skip the longer sentences early in training, with a length threshold that would gradually increase depending on the `global_step`. _Second question_: Could you reserve one batch id (e.g. -1) in `group_by_window` to tag the batches that will be skipped ? At the moment, it seems that all the ids are considered, even the negative values, and it would not be restrictive at all to allow only positive values (as there would still be 63 bits left to group the inputs). Thus, in `key_func` we could simply compare the input length with the threshold and return a negative value when it is above it.

Apologies if both functionalities are already available, feel free to stackoverflow me.

**Comment by jart** [comment]
> feel free to stackoverflow me

I haven't heard that one before. [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) is probably your best bet when there's any uncertainty about whether or not a feature is necessary. We triagers can also take care of the work of tagging folks on the team for you in the future. I will note that you made me smile, so I'm going to try my best to fill in for the community.

Speaking as someone who only recently began learning `contrib.data`, from what I understand these are stateful iterative ops, so I'm not sure if the sort of sorting you described would be possible. Have you considered collecting all the data and using [tf.nn.top_k](https://www.tensorflow.org/api_docs/python/tf/nn/top_k)? I'm also assuming you're also familiar with `tf.cond` which has been traditionally used to implement skipping type logic.

If those things don't help, please give StackOverflow a try. I'll be happy to reopen this issue if provided new information.

**Comment by georgesterpu** [comment]
You are completely right, @jart. It would be more appropriate if I rephrase the last sentence as : "After going through the entire support thread in [[1]](https://github.com/tensorflow/tensorflow/issues/7951) and searching related content on StackOverflow too, to the best of my knowledge these functionalities have not been implemented yet." 

If you are unsure about the possibility of sorting a Dataset after applying the `group_by_window()` transformation, perhaps it would be fair to re-open the issue. Could you also help me find out what is the implicit order of the batches when this transformation is applied, please ?

On batch skipping, could you please provide me an example of ignoring the batches grouped by a range of ids within `group_by_window` ? I am probably overthinking this, yet it seems that the id returned by `key_func` would not be available anymore at runtime. Maybe `reduce_func` could discard some batches instead ?

Thank you

**Comment by jart** [comment]
Just so there's no misunderstanding, our team is always grateful when members of the community take the time to leave us feedback. My goal is to be friendly and helpful in this process. Per your request, I'm happy to reopen this issue, so the next triager can take a look in a day or so.

**Comment by georgesterpu** [comment]
Thanks, Justine. One workaround that crossed my mind is storing the inputs in separate TFRecords based on their lengths and concatenating several Datasets when creating the iterator. Having this feature on the iterator would only make things more compact.

**Comment by fanlu** [comment]
I also meet this problem. And more I want to use distribute tf to train my deepspeech2 model,So It requests dataset.get_next() to yield the same bucket_id data in distributed machines.I also try dataset.shard(), but this function returned dataset's bucket_id is not same in one batch.How do I implement this function?

**Comment by CFAndy** [comment]
It will be helpful to provide this for distributed training with MPI based framework.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by georgesterpu** [comment]
Hello, as a future reference, skipping examples longer than some `max_length` could be done like this:

`dataset = dataset.filter(lambda elem: tf.shape(elem)[0] < max_length)`
